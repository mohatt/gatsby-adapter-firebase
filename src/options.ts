import { Joi, PluginOptionsSchemaJoi, validateOptionsSchema } from 'gatsby-plugin-utils'
import type { BuildHostingOptions } from './lib/build-hosting.js'
import type { BuildFunctionsOptions } from './lib/build-functions.js'

export interface AdapterOptions extends Partial<BuildHostingOptions & BuildFunctionsOptions> {
  excludeDatastoreFromEngineFunction?: boolean
}

export type ValidatedAdapterOptions = Readonly<Required<AdapterOptions>>

export type ValidatedAdapterResult =
  | { options: ValidatedAdapterOptions; warnings?: string[] }
  | { errors: string[] }

/**
 * Defines the schema for adaptor options.
 */
export const getOptionsSchema = (joi: PluginOptionsSchemaJoi) => {
  return joi
    .object({
      hostingTarget: joi
        .string()
        .description('Firebase Hosting target name to update within firebase.json')
        .default('gatsby'),
      functionsConfig: joi
        .object()
        .unknown(true)
        .description('Default firebase-functions HTTPS options applied to every generated function')
        .optional(),
      functionsConfigOverride: joi
        .object()
        .pattern(
          joi.string(),
          joi.object().unknown(true).description('HTTPS options override for a single function'),
        )
        .description(
          'Per-function overrides keyed by original Gatsby functionId, merged on top of functionsConfig',
        )
        .default({}),
      functionsOutDir: joi
        .string()
        .description(
          'Directory (relative to project root) where the adapter writes the functions codebase',
        )
        .default('.firebase/functions'),
      functionsCodebase: joi
        .string()
        .description('Firebase codebase name used when registering the generated functions bundle')
        .default('gatsby'),
      functionsRuntime: joi
        .string()
        .valid('nodejs20', 'nodejs22')
        .description('Runtime identifier (e.g. nodejs20) used for generated Firebase functions')
        .default('nodejs20'),
      storageBucket: joi
        .string()
        .description(
          'Cloud Storage bucket name used for cached responses (defaults to project bucket)',
        ),
      excludeDatastoreFromEngineFunction: joi
        .boolean()
        .description('Skip bundling Gatsby LMDB datastore in the functions used for SSR/DSG')
        .default(false),
    })
    .label('options')
}

export const validateOptions = async (options: AdapterOptions): Promise<ValidatedAdapterResult> => {
  try {
    const { value, warning } = await validateOptionsSchema(
      getOptionsSchema(Joi),
      options as Record<string, unknown>,
    )
    return {
      options: value as ValidatedAdapterOptions,
      warnings: warning?.details.length ? warning.details.map((w) => w.message) : undefined,
    }
  } catch (err) {
    return {
      errors: Array.isArray(err.details)
        ? (err.details as Array<{ message: string }>).map((e) => e.message)
        : [String(err)],
    }
  }
}
