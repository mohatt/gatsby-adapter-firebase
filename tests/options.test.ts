import { describe, expect, it } from 'vitest'
import { AdapterOptions, validateOptions } from '../src/options.js'

interface TestCase {
  title: string
  options: AdapterOptions | Record<string, unknown>
}

const testCases: TestCase[] = [
  {
    title: 'accepts defaults with no options',
    options: {},
  },
  {
    title: 'accepts valid options',
    options: {
      hostingTarget: 'custom',
      functionsOutDir: 'custom/functions',
      functionsCodebase: 'custom',
      functionsRuntime: 'nodejs22',
      functionsConfig: {
        region: 'us-west1',
        timeoutSeconds: 120,
        unsupportedCustomFlag: { unknown: 'value' },
      },
      functionsConfigOverride: {
        'ssr-engine': {
          region: 'asia-northeast1',
          concurrency: 10,
        },
      },
      storageBucket: 'test-project.appspot.com',
      excludeDatastoreFromEngineFunction: true,
    },
  },
  {
    title: 'warns unknown options',
    options: {
      unknown: 'value',
    },
  },
  {
    title: 'rejects invalid options',
    options: {
      hostingTarget: 123,
      functionsOutDir: false,
      functionsCodebase: ['invalid'],
      functionsRuntime: 'nodejs18',
      functionsConfig: 'invalid',
      functionsConfigOverride: {
        'ssr-engine': 'invalid',
      },
      storageBucket: [123],
      excludeDatastoreFromEngineFunction: 'notBoolean',
    },
  },
  {
    title: 'rejects empty options',
    options: {
      hostingTarget: '',
      functionsOutDir: '',
      functionsCodebase: '',
      functionsRuntime: '',
      storageBucket: '',
    },
  },
  {
    title: 'rejects invalid options object',
    options: null,
  },
]

describe('options', () => {
  it.each(testCases)('$title', async ({ options }) => {
    await expect(validateOptions(options)).resolves.toMatchSnapshot('result')
  })
})
