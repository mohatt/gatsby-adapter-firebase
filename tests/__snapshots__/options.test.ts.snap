// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options > 'accepts defaults with no options' > result 1`] = `
{
  "options": {
    "excludeDatastoreFromEngineFunction": false,
    "functionsCodebase": "gatsby",
    "functionsConfigOverride": {},
    "functionsOutDir": ".firebase/functions",
    "functionsRuntime": "nodejs20",
    "hostingTarget": "gatsby",
  },
  "warnings": undefined,
}
`;

exports[`options > 'accepts valid options' > result 1`] = `
{
  "options": {
    "excludeDatastoreFromEngineFunction": true,
    "functionsCodebase": "custom",
    "functionsConfig": {
      "region": "us-west1",
      "timeoutSeconds": 120,
      "unsupportedCustomFlag": {
        "unknown": "value",
      },
    },
    "functionsConfigOverride": {
      "ssr-engine": {
        "concurrency": 10,
        "region": "asia-northeast1",
      },
    },
    "functionsOutDir": "custom/functions",
    "functionsRuntime": "nodejs22",
    "hostingTarget": "custom",
    "storageBucket": "test-project.appspot.com",
  },
  "warnings": undefined,
}
`;

exports[`options > 'rejects empty options' > result 1`] = `
{
  "errors": [
    ""hostingTarget" is not allowed to be empty",
    ""functionsOutDir" is not allowed to be empty",
    ""functionsCodebase" is not allowed to be empty",
    ""functionsRuntime" must be one of [nodejs20, nodejs22]",
    ""functionsRuntime" is not allowed to be empty",
    ""storageBucket" is not allowed to be empty",
  ],
}
`;

exports[`options > 'rejects invalid options object' > result 1`] = `
{
  "errors": [
    ""options" must be of type object",
  ],
}
`;

exports[`options > 'rejects invalid options' > result 1`] = `
{
  "errors": [
    ""hostingTarget" must be a string",
    ""functionsConfig" must be of type object",
    ""functionsConfigOverride.ssr-engine" must be of type object",
    ""functionsOutDir" must be a string",
    ""functionsCodebase" must be a string",
    ""functionsRuntime" must be one of [nodejs20, nodejs22]",
    ""storageBucket" must be a string",
    ""excludeDatastoreFromEngineFunction" must be a boolean",
  ],
}
`;

exports[`options > 'warns unknown options' > result 1`] = `
{
  "options": {
    "excludeDatastoreFromEngineFunction": false,
    "functionsCodebase": "gatsby",
    "functionsConfigOverride": {},
    "functionsOutDir": ".firebase/functions",
    "functionsRuntime": "nodejs20",
    "hostingTarget": "gatsby",
    "unknown": "value",
  },
  "warnings": [
    ""unknown" is not allowed",
  ],
}
`;
